### PR TITLE
Lint, test, fix mongo helper partial permissions

### PR DESCRIPTION
**Pull request by bufke**
## Notes

This PR should contain no user facing changes. We need to rewrite a sizable portion of the affected files to support AND and OR in partial permissions. To make changes manageable, I split up the PR. This PR contains only lint, minor bug fixes, tests, and type corrections. All things that are only tangentially related to partial permissions. The partial permissions, partial rewrite is complex and functional types are key to successful implementation. Getting the linting done here will greatly simplify the job while keeping our PR diffs very readable.

-  Run linter on mongo helper and partial perm files
-  Add mypy and ruff pyproject settings
-  Refactor mongo helper default args, fix documented types
-  Add mongo helper unit test
-  Add partial perm unit tests (not all pass here, marked to not run)
-  Improve validation on asset permission assignment

While the majority of the changes are cosmetic, it contains some fixes and simplifications which could in theory introduce bugs that we don't have test coverage on.

Testing done: I created some shared partial permission in the existing UI and saw no error.

## Related issues

- Part of [TASK-402](https://www.notion.so/kobotoolbox/And-or-support-for-mongo-helper-and-partial-permissions-API-14c82b9c0d5d45c98bd3d011caa50b90)
- Partial replacement for https://github.com/kobotoolbox/kpi/pull/4757
- Replaces https://github.com/kobotoolbox/kpi/pull/4718 (close when merged)

### CODE REVIEW

File: kpi/serializers/v2/asset_permission_assignment.py [hunk 1]
```diff
@@ -65,9 +64,7 @@ def create(self, validated_data):
         user = validated_data['user']
         asset = validated_data['asset']
         if asset.owner_id == user.id:
-            raise serializers.ValidationError(
-                {'user': t(ASSIGN_OWNER_ERROR_MESSAGE)}
-            )
+            raise serializers.ValidationError({'user': t(ASSIGN_OWNER_ERROR_MESSAGE)})
```
Comment (noliveleger) [inline kpi/serializers/v2/asset_permission_assignment.py:67]: It's going over the 80 characters limit (but below 90*). Did you do this in purpose or your linter is configured this way? 

`*` Old Kobo Python coding convention discussion: We usually break at 80 but allow up to 90 when it's only for few characters.

File: kpi/serializers/v2/asset_permission_assignment.py [hunk 2]
```diff
@@ -27,24 +28,22 @@
 )
 from kpi.utils.urls import absolute_resolve
 
-
 ASSIGN_OWNER_ERROR_MESSAGE = "Owner's permissions cannot be assigned explicitly"
 
 
 class AssetPermissionAssignmentSerializer(serializers.ModelSerializer):
-
```
Comment (noliveleger) [inline kpi/serializers/v2/asset_permission_assignment.py:35]: 💔 . Your linter is going to be crazy on all my code. I do add an extra line in purpose.

File: kpi/serializers/v2/asset_permission_assignment.py [hunk 3]
```diff
@@ -275,13 +266,14 @@ def __get_permission_hyperlink(self, codename):
...
-                       request=self.context.get('request', None))
+        return reverse(
+            'permission-detail',
+            args=(codename,),
+            request=self.context.get('request', None),
+        )
 
 
 class PartialPermissionField(serializers.Field):
-
```
Comment (noliveleger) [inline kpi/serializers/v2/asset_permission_assignment.py:284]: 💔  again.

File: kpi/serializers/v2/asset_permission_assignment.py [hunk 4]
```diff
@@ -318,11 +309,13 @@ class AssetBulkInsertPermissionSerializer(serializers.Serializer):
     Warning: If less queries are sent to DB, it consumes more CPU and memory.
     The bigger the assignments are, the bigger the resources footprint will be.
     """
+
     assignments = serializers.ListField(child=PermissionAssignmentSerializer())
 
     @dataclass(frozen=True)
     class PermissionAssignment:
-        """ A more-explicit alternative to a simple tuple """
+        """A more-explicit alternative to a simple tuple"""
```
Comment (noliveleger) [inline kpi/serializers/v2/asset_permission_assignment.py:317]: Your linter really don't like when the code breaths a little :-P 
I would leave the extra space alone or even better I would put the docstrings on separated  lines.

```python
"""
A more-explicit alternative to a simple tuple
"""
```

I prefer this way even if the sentence is short. Moreover if there is a diff at a later time, it does not involve the triple quotes.

File: kpi/serializers/v2/asset_permission_assignment.py [hunk 5]
```diff
@@ -430,9 +421,9 @@ def get_set_of_incoming_assignments(
             # always require a fully-fledged `User` object? Until then, keep a
             # stupid object cache thing because `assign_perm()` and
             # `remove_perm()` REQUIRE user objects
-            user_pk_to_obj_cache[
-                incoming_assignment['user'].pk
-            ] = incoming_assignment['user']
+            user_pk_to_obj_cache[incoming_assignment['user'].pk] = incoming_assignment[
+                'user'
+            ]
```
Comment (noliveleger) [inline kpi/serializers/v2/asset_permission_assignment.py:426]: Same comment as https://github.com/kobotoolbox/kpi/pull/4776#discussion_r1434101171

File: kpi/tests/api/v2/test_api_asset_permission_assignment.py [hunk 6]
```diff
@@ -1,30 +1,29 @@
 # coding: utf-8
 from copy import deepcopy
 
-from django.contrib.auth.models import User, Permission
+from django.contrib.auth.models import Permission, User
```
Comment (noliveleger) [inline kpi/tests/api/v2/test_api_asset_permission_assignment.py:4]: ❤️

File: kpi/serializers/v2/asset_permission_assignment.py (hunk 2 above)
Comment (noliveleger) [reply kpi/serializers/v2/asset_permission_assignment.py:35]: It happens several times in this PR. I would keep the extra line at the beginning of classes (and methods).

File: kpi/tests/api/v2/test_api_asset_permission_assignment.py [hunk 7]
```diff
@@ -659,3 +634,209 @@ def test_no_assignments_saved_on_error(self):
         # than the one they already had, i.e.: 'view_submissions'.
         self.assertFalse(self.asset.has_perm(self.someuser, PERM_DELETE_SUBMISSIONS))
         self.assertFalse(self.asset.has_perm(self.anotheruser, PERM_CHANGE_SUBMISSIONS))
+
+    # TODO
```
Comment (noliveleger) [inline kpi/tests/api/v2/test_api_asset_permission_assignment.py:638]: We could also use `@pytest.mark.skip(reason=...)`

File: kpi/tests/api/v2/test_api_asset_permission_assignment.py [hunk 8]
```diff
@@ -659,3 +634,209 @@ def test_no_assignments_saved_on_error(self):
...
+                'filters': [
+                    [{'my_question': 'my_response1'}, {'my_question2': 'my_response2'}],
+                ],
+            },
+        ]
+
+        assert expected == returned_partial_perms
+
+    def test_partial_permission_invalid(self):
+        """Poorly formatted json example"""
```
Comment (noliveleger) [inline kpi/tests/api/v2/test_api_asset_permission_assignment.py:821]: 3 lines please? 🥹

File: pyproject.toml [hunk 9]
```diff
@@ -8,3 +8,19 @@ skip-string-normalization = true
 profile = "black"
 known_first_party = "kobo"
 no_lines_before = ["LOCALFOLDER"]
+
+[tool.ruff.format]
+quote-style = "single"
+[tool.ruff.lint]
+extend-select = ["I"]
```
Comment (noliveleger) [inline pyproject.toml:15]: I would add a comment to explain what's the purpose of this line.

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kpi/serializers/v2/asset_permission_assignment.py (hunk 1 above)
Comment (bufke) [reply kpi/serializers/v2/asset_permission_assignment.py:67]: It appears screwed up my local config to use default 88 when it was already set to 80. Using the formatter with 80 is OK right? I will update. Note the formatter greatly helps in the (soon to be) large permission dict (") and json (') that we need in docs and tests.

**Comment by bufke** [comment]
Important - squash and merge this otherwise we the commit history will be contain contradictory changes that will make understanding history harder.

**Review by noliveleger — APPROVED** [review]

**Pull request merged**